File: evetrade_functions/esi_handlers/get_orders/get-hauling-orders.js
// Compares profitable hailing orders between stations and/or regions (or within the same region)
const AWS = require('aws-sdk');
const https = require('https');
const MarketData = require('./helper_modules/MarketData.js').MarketData;

AWS.config.update({region: 'us-east-1'});
const s3 = new AWS.S3();

let typeIDToName, stationIdToName, systemIdToSecurity;

const jumpCount = {};

/**
 * Generates and executes market data requests based on the requested queries
 * @param {*} inputList 
 * @param {*} orderType 
 * @returns Market Data Mapping which has all the requests executed
 */
function aggregate_mapping(locations, orderType) {
    const marketDataMapping = {};
    
    for (const region in locations){
        const stations = locations[region];
        
        marketDataMapping[region] = new MarketData(region, orderType, stations);
        marketDataMapping[region].executeRequest();
    }
    
    return marketDataMapping;
}

/**
 * Once all orders are received generate aggregate maps
 * @param {*} fromMapping The current mapping of MarketData classes for initial orders
 * @param {*} toMapping  The current mapping of MarketData classes for closing orders
 * @returns Mapping which aggregates all region orders
 */
function get_aggregate_orders(fromMapping, toMapping) {
    return new Promise(function (resolve) {
        const interval = setInterval(async function() {
            let completeExecution = true;
            
            for (const region in fromMapping){
                if (!fromMapping[region].completeExecution) {
                    completeExecution = false;
                }
            }
            
            for (const region in toMapping){
                if (!toMapping[region].completeExecution) {
                    completeExecution = false;
                }
            }
            
            if (completeExecution) {
                clearInterval(interval);
                
                const orders = {
                    'from': [],
                    'to': []
                };
                
                for (const region in fromMapping){
                    orders['from'] = orders['from'].concat(fromMapping[region].orders);
                }

                for (const region in toMapping){
                    orders['to'] = orders['to'].concat(toMapping[region].orders);
                }

                resolve(orders);
            }
            

        });
    });
}

/**
 * Maps to cheaper/expensive by station to ensure one item per station
 * @param {*} orders The full list of orders
 * @param {*} cheaper Whether we want cheaper (sell orders) or more expensive (buy orders) 
 * @returns Mapping of inidividual items in given stations that are most adventageous
 */
function remap_orders(orders, cheaper) {
    const newOrders = {};
    for (const obj of orders) {
        const typeId = obj.type_id;
        const stationId = obj.location_id;
        if (newOrders[typeId] === undefined) {
            newOrders[typeId] = {};
        } 

        if(newOrders[typeId][stationId] === undefined) {
            newOrders[typeId][stationId] = obj;
        } else {
            if (cheaper && obj.price < newOrders[typeId][stationId].price) {
                newOrders[typeId][stationId] = obj;
            } else if (!cheaper && obj.price > newOrders[typeId][stationId].price) {
                newOrders[typeId][stationId] = obj;
            }
        }
    }
    return newOrders;
}

/**
 * Removes the type IDs that do not align between FROM and TO orders
 * @param {*} fromOrders Orders at the initiating end
 * @param {*} toOrders Orders at the closing end
 * @returns Subset of all the Ids (should be equal in size)
 */
function remove_mismatch_type_ids(fromOrders, toOrders) {
    const fromIds = Object.keys(fromOrders);
    const toIds = Object.keys(toOrders);

    console.log(`Before: From Order Count = ${fromIds.length} and To Order Count = ${toIds.length}`);
    
    for (const id of fromIds) {
        if(toOrders[id] === undefined) {
            delete fromOrders[id];
        }
    }
    
    for (const id of toIds) {
        if(fromOrders[id] === undefined) {
            delete toOrders[id];
        }
    }
    
    console.log(`After: From Order Count = ${Object.keys(fromOrders).length} and To Order Count = ${Object.keys(toOrders).length}`);

    return {
        'from': fromOrders,
        'to': toOrders
    };
}

/**
 * Based on given parameters it returns a set of valid trades which meet initial parameters
 * @param {*} fromOrders Originating orders
 * @param {*} toOrders Closing orders
 * @param {*} tax Tax rate
 * @param {*} minProfit Minimum Profit
 * @param {*} minROI Minimum ROI 
 * @param {*} maxBudget Maximum Budget
 * @param {*} maxWeight Maximum Weight
 * @returns Map of valid trades
 */
function get_valid_trades(fromOrders, toOrders, tax, minProfit, minROI, maxBudget, maxWeight) {
    const ids = Object.keys(fromOrders);
    const validTrades = [];

    for (const id of ids) {
        const fromStations = fromOrders[id];
        const toStations = toOrders[id];

        const fromKeys = Object.keys(fromStations);
        for (const fromStation of fromKeys) {
            const initialOrder = fromOrders[id][fromStation];

            const toKeys = Object.keys(toStations);
            for (const toStation of toKeys) {
                const closingOrder = toOrders[id][toStation];
                const volume = closingOrder.volume_remain < initialOrder.volume_remain ? closingOrder.volume_remain : initialOrder.volume_remain;

                const initialPrice = initialOrder.price * volume;
                const salePrice = closingOrder.price * volume * (1-tax);
                const profit = salePrice - initialPrice;
                const ROI = (salePrice-initialPrice)/initialPrice;
                const weight = typeIDToName[initialOrder.type_id].volume * volume;
                if (profit > minProfit && ROI >= minROI && initialPrice <= maxBudget && weight < maxWeight) {
                    const newRecord = {
                        'Item': typeIDToName[initialOrder.type_id].name,
                        'From': {
                            'name': stationIdToName[initialOrder.location_id],
                            'system_id': initialOrder.system_id,
                            'rating': systemIdToSecurity[initialOrder.system_id]["rating"],
                            'security_code': systemIdToSecurity[initialOrder.system_id]["security_code"]
                        },
                        'Quantity': volume,
                        'Buy Price': initialOrder.price,
                        'Net Costs': volume * initialOrder.price,
                        'Take To': {
                            'name': stationIdToName[closingOrder.location_id],
                            'system_id': closingOrder.system_id,
                            'rating': systemIdToSecurity[closingOrder.system_id]["rating"],
                            'security_code': systemIdToSecurity[closingOrder.system_id]["security_code"]
                        },
                        'Sell Price': closingOrder.price,
                        'Net Sales': volume * closingOrder.price,
                        'Gross Margin': volume * (closingOrder.price - initialOrder.price),
                        'Sales Taxes': volume * (closingOrder.price * tax / 100),
                        'Net Profit': profit,
                        'R.O.I.': (100 * ROI).toFixed(2) + "%",
                        'Total Volume (m3)': weight,
                    };

                    validTrades.push(newRecord);

                    jumpCount[`${initialOrder.system_id}-${closingOrder.system_id}`] = '';
                }
            }
        }
    }
    return validTrades;
}

let request_complete, request_count;

async function get_number_of_jumps(safety, validTrades) {
    const routes = Object.keys(jumpCount);
    request_count = 0;
    request_complete = 0;

    for (const route of routes) {
        const fromSystem = route.split('-')[0];
        const toSystem = route.split('-')[1];
        const url = `https://esi.evetech.net/latest/route/${fromSystem}/${toSystem}/?datasource=tranquility&flag=${safety}`;
        request_count += 1;

        new Promise((resolve, reject) => {
            const req = https.get(url, {}, res => {
                let rawData = '';
                
                res.on('data', chunk => {
                    rawData += chunk;
                });
                
                res.on('end', () => {
                    try {
                        request_complete += 1;
                        resolve(JSON.parse(rawData));
                    } catch (err) {
                        reject(new Error(err));
                    }
                });
            });
            
            req.on('error', err => {
                reject(new Error(err));
            });
        }).then(function(data){
            jumpCount[`${data[0]}-${data[data.length-1]}`] = data.length - 1;
            
        });
    }
    
    return new Promise(function (resolve) {
        const interval = setInterval(async function() {
            if (request_count == request_complete) {
                clearInterval(interval);
                
                for (const trades of validTrades) {
                    const fromSystem = trades['From']['system_id'];
                    const toSystem = trades['Take To']['system_id'];
                    trades['Jumps'] = jumpCount[`${fromSystem}-${toSystem}`];
                    trades['Profit per Jump'] = (parseFloat(trades['Net Profit']) / parseInt(trades['Jumps'], 10)).toFixed(2);
                }
                
                resolve(validTrades);
            }
        });
    });
}


function get_mappings() {
    const DOWNLOAD_PARAMS = {
        Bucket: 'evetrade',
        Key: `resources/typeIDToName.json`
    };

    s3.getObject(DOWNLOAD_PARAMS).promise()
    .then( data => { 
        console.log(`Successfully retrieved typeIDToName`);
        typeIDToName = JSON.parse(data.Body.toString('utf-8'));
    })
    .catch( err => { console.log(`Failed`, err); 
    });

    DOWNLOAD_PARAMS.Key = `resources/stationIdToName.json`;
    s3.getObject(DOWNLOAD_PARAMS)
    .promise()
    .then( data => { 
        console.log(`Successfully retrieved stationIdToName`);
        stationIdToName = JSON.parse(data.Body.toString('utf-8'));
    })
    .catch( err => { 
        console.log(`Failed`, err); 
    });

    DOWNLOAD_PARAMS.Key = `resources/systemIdToSecurity.json`;
    s3.getObject(DOWNLOAD_PARAMS)
    .promise()
    .then( data => { 
        console.log(`Successfully retrieved systemIdToSecurity`);
        systemIdToSecurity = JSON.parse(data.Body.toString('utf-8'));
    })
    .catch( err => { 
        console.log(`Failed`, err); 
    });
}

function convert_locations(locations) {
    const splitValues = {};
    
    const individualLocations = locations.split(',');
    
    for (const location of individualLocations) {
        if (location.indexOf(':') > 0) {
            const region = parseInt(location.split(':')[0], 10);
            const station = parseInt(location.split(':')[1], 10);
            
            if(splitValues[region] === undefined) {
                splitValues[region] = [];
            }
            
            splitValues[region].push(station);
        } else {
            splitValues[location] = [];
        }
    }
    
    return splitValues;
}

/**
 * Lambda function handler
 * @param {*} event 
 * @param {*} context 
 * @returns Payload of profitable trades
 */
exports.handler = async function(event, context) {
    console.log(event);
    const startTime = new Date();
    const queries = event["queryStringParameters"];
    const AGG_FROM = aggregate_mapping(convert_locations(queries['from']), 'sell');
    const AGG_TO = aggregate_mapping(convert_locations(queries['to']), 'buy');
    const SALES_TAX = queries['tax'] === undefined ? 0.08 : queries['tax'];
    const MIN_PROFIT = queries['minProfit'] === undefined ? 500000 : queries['minProfit'];
    const MIN_ROI = queries['minROI'] === undefined ? 0.04 : queries['minROI'];
    const MAX_BUDGET = queries['maxBudget'] === undefined ? Number.MAX_SAFE_INTEGER : queries['maxBudget'];
    const MAX_WEIGHT = queries['maxWeight'] === undefined ? Number.MAX_SAFE_INTEGER : queries['maxWeight'];
    const ROUTE_SAFETY = queries['routeSafety'] === undefined ? 'shortest' : queries['routeSafety'];
    const FROM_TYPE = queries['fromType'] === undefined ? 'sell' : queries['fromType'];
    const TO_TYPE = queries['toType'] === undefined ? 'buy' : queries['toType'];
    
    // Get cached mappings files for easier processing later.
    get_mappings();

    console.log(`Mapping retrieval took: ${(new Date() - startTime) / 1000} seconds to process.`);

    let orders = await get_aggregate_orders(AGG_FROM, AGG_TO);
    
    // Grab one item per station in each each (cheaper for sell orders, expensive for buy orders)
    // Remove type Ids that do not exist in each side of the trade
    orders = remove_mismatch_type_ids(
        remap_orders(orders['from'], FROM_TYPE == 'buy' ? false : true), 
        remap_orders(orders['to'], TO_TYPE == 'buy' ? false : true)
    );

    let validTrades = get_valid_trades(orders['from'], orders['to'], SALES_TAX, MIN_PROFIT, MIN_ROI, MAX_BUDGET, MAX_WEIGHT);
    console.log(`Valid Trades = ${validTrades.length}`);

    validTrades = await get_number_of_jumps(ROUTE_SAFETY, validTrades);

    console.log(`Full analysis took: ${(new Date() - startTime) / 1000} seconds to process.`);

    return {
        'body': JSON.stringify(validTrades)
    };
};

File: evetrade_functions/esi_handlers/get_orders/get_hauling_orders.py
'''
Serverless Lambda Function which provides profitable
hauling orders for given a given API Request
'''

import sys
import time
import json
import asyncio
import math
import aiohttp
import boto3
from market_data import MarketData

jump_count = {}

def round_up(number):
    '''
    Rounds a value up to the nearest 2 decimal places
    '''
    return math.ceil(number*100)/100

def convert_locations(locations, order_type):
    '''
    Converts plaintext string locations and orders into market_data classes
    '''
    individual_locations = locations.split(',')
    converted = {}

    for location in individual_locations:
        if ':' in location:
            split = location.split(':')
            region = int(split[0])
            station = int(split[1])

            if region not in converted:
                converted[region] = []

            converted[region].append(station)
        else:
            converted[location] = []

    market_requests = []

    for region, station_ids in converted.items():
        market_requests.append(MarketData(region, order_type, station_ids))

    return market_requests


async def get_order_data(market_requests, chunk_size):
    '''
    Asynchronously retrives order data for all market data requests
    by chunking the requests accordingly to avoid overworking any
    network bandwidth for the larger queries
    '''
    chunk_requests = [
        market_requests[i:i + chunk_size] for i in range(0, len(market_requests), chunk_size)
    ]

    chunk_count = len(chunk_requests)
    print(f'Processing in {chunk_count} chunks')

    orders = []

    for chunk_idx, chunk in enumerate(chunk_requests):
        tasks = []
        print(f'-- Chunk: #{chunk_idx+1}')
        for market_data in chunk:
            tasks.append(
                asyncio.ensure_future(market_data.execute_requests())
            )

        all_orders = await asyncio.gather(*tasks)

        for order_page in all_orders:
            orders = orders + order_page

    return orders

async def async_order_request(from_locations, from_size, to_locations, to_size):
    '''
    Asynchronously gets all the orders related to the from and to locations
    '''
    tasks = [
        asyncio.ensure_future(initialize_mapping()),
        asyncio.ensure_future(get_order_data(from_locations, from_size)),
        asyncio.ensure_future(get_order_data(to_locations, to_size))
    ]

    data = await asyncio.gather(*tasks)

    return {
        'mappings': data[0],
        'from': data[1],
        'to': data[2]
    }

async def initialize_mapping():
    '''
    Grabs static mapping files that are retrieved from EVE SDE
    and stored on S3 for rapid calculations
    '''
    s3_client = boto3.client(
        's3'
    )

    mapping_data = {
        'typeIDToName': {},
        'stationIdToName': {},
        'systemIdToSecurity': {}
    }

    for s3_file in mapping_data.copy():
        obj = s3_client.get_object(Bucket = 'evetrade', Key = f'resources/{s3_file}.json')
        data = obj['Body'].read().decode('utf-8')
        mapping_data[s3_file] = json.loads(data)

        print(f'- Successfully loaded {s3_file}.json -')

    return mapping_data


def remap_orders(orders, cheaper):
    '''
    Maps to cheaper / expensive by station to ensure one item per station
    '''

    new_orders = {}
    for order in orders:
        type_id = order['type_id']
        station_id = order['location_id']

        # TODO find how to get Citadel Name mapping
        if station_id > 99999999:
            continue

        if type_id not in new_orders:
            new_orders[type_id] = {}

        if station_id not in new_orders[type_id]:
            new_orders[type_id][station_id] = order
        else:
            if order['volume_remain'] > 0: # Bug where API returns orders with 0 volume
                if cheaper and order['price'] < new_orders[type_id][station_id]['price']:
                    new_orders[type_id][station_id] = order
                elif not cheaper and order['price'] > new_orders[type_id][station_id]['price']:
                    new_orders[type_id][station_id] = order

    return new_orders


def remove_mismatch_type_ids(orders):
    '''
    Removes the type IDs that do not align between FROM and TO orders
    '''

    from_keys = list(orders['from'].keys())

    for type_id in from_keys:
        if type_id not in orders['to']:
            del orders['from'][type_id]

    to_keys = list(orders['to'].keys())

    for type_id in to_keys:
        if type_id not in orders['from']:
            del orders['to'][type_id]

    return orders


def get_valid_trades(data, tax, min_profit, min_roi, max_budget, max_weight):
    '''
    Based on given parameters it returns a set of valid trades which meet initial parameters
    '''

    mappings = data['mappings']
    type_id_to_name = mappings['typeIDToName']
    station_id_to_name = mappings['stationIdToName']
    system_id_to_security = mappings['systemIdToSecurity']

    from_orders = data['from']
    to_orders = data['to']

    valid_trades = []
    type_ids = list(from_orders.keys())

    for type_id in type_ids:
        from_stations = from_orders[type_id]
        to_stations = to_orders[type_id]

        from_keys = list(from_stations.keys())
        for from_station in from_keys:
            initial_order = from_orders[type_id][from_station]

            to_keys = list(to_stations.keys())
            for to_station in to_keys:
                closing_order = to_orders[type_id][to_station]
                volume = closing_order['volume_remain'] \
                    if closing_order['volume_remain'] < initial_order['volume_remain'] \
                    else initial_order['volume_remain']

                initial_price = initial_order['price'] * volume
                sale_price = closing_order['price'] * volume * (1-tax)
                profit = sale_price - initial_price

                return_on_investment = (sale_price - initial_price) / initial_price
                weight = type_id_to_name[
                        str(initial_order['type_id'])
                    ]['volume'] * volume
                if (
                    profit > min_profit
                        and return_on_investment >= min_roi
                        and initial_price <= max_budget
                        and weight < max_weight
                    ):
                    new_record = {
                        'Item': type_id_to_name[
                                str(initial_order['type_id'])
                            ]['name'],
                        'From': {
                            'name': station_id_to_name[
                                    str(initial_order['location_id'])
                                ],
                            'system_id': initial_order['system_id'],
                            'rating': system_id_to_security[
                                    str(initial_order['system_id'])
                                ]['rating'],
                            'security_code': system_id_to_security[
                                    str(initial_order['system_id'])
                                ]['security_code']
                        },
                        'Quantity': volume,
                        'Buy Price': initial_order['price'],
                        'Net Costs': volume * initial_order['price'],
                        'Take To': {
                            'name': station_id_to_name[
                                    str(closing_order['location_id'])
                                ],
                            'system_id': closing_order['system_id'],
                            'rating': system_id_to_security[
                                    str(closing_order['system_id'])
                                ]['rating'],
                            'security_code': system_id_to_security[
                                    str(closing_order['system_id'])
                                ]['security_code']
                        },
                        'Sell Price': round_up(closing_order['price']),
                        'Net Sales': round_up(volume * closing_order['price']),
                        'Gross Margin': round_up(
                                volume * (closing_order['price'] - initial_order['price'])
                            ),
                        'Sales Taxes': round_up(volume * (closing_order['price'] * tax / 100)),
                        'Net Profit': round_up(profit),
                        'R.O.I.': f'{round_up(100 * return_on_investment)}%',
                        'Total Volume (m3)': round_up(weight),
                    }

                    valid_trades.append(new_record)

                    jump_count[f'{initial_order["system_id"]}-{closing_order["system_id"]}'] = ''

    return valid_trades

async def get_jump_count(session, from_system, to_system, route_safety):
    '''
    Gets jump count for a particular system to system route
    '''
    base_url = 'https://esi.evetech.net/latest/route/'
    url = f'{base_url}{from_system}/{to_system}/?datasource=tranquility&flag={route_safety}'

    async with session.get(url) as resp:
        data = await resp.json()
        jump_count[f'{from_system}-{to_system}'] =  len(data) - 1

async def get_all_jump_counts(trades, route_safety):
    '''
    Gets all jump counts needed for viable trades
    '''
    async with aiohttp.ClientSession() as session:
        tasks = []
        for routes in jump_count:
            route = routes.split('-')
            tasks.append(
                asyncio.ensure_future(get_jump_count(session, route[0], route[1], route_safety))
            )

        await asyncio.gather(*tasks)

    for trade in trades:
        from_system = trade['From']['system_id']
        to_system = trade['Take To']['system_id']
        trade['Jumps'] = jump_count[f'{from_system}-{to_system}']
        trade['Profit per Jump'] = round_up(float(trade['Net Profit']) / int(trade['Jumps']))

    return trades

def lambda_handler(event, context):
    '''
    Serverless Lambda Handler which is invoked at initialization
    and event body is passed which contains request information from API Gateway
    '''
    print(event, context)

    start_time = time.time()

    queries = event['queryStringParameters']

    sales_tax = 0.08 if 'tax' not in queries else queries['tax']
    min_profit = 500000 if 'minProfit' not in queries else queries['minProfit']
    min_roi = 0.04 if 'minROI' not in queries else queries['minROI']
    max_budget = sys.maxsize if 'maxBudget' not in queries else queries['maxBudget']
    max_weight = sys.maxsize if 'maxWeight' not in queries else queries['maxWeight']
    route_safety = 'shortest' if 'routeSafety' not in queries else queries['routeSafety']
    from_type = 'sell' if 'fromType' not in queries else queries['fromType']
    to_type = 'buy' if 'toType' not in queries else queries['toType']

    chunk_size = 10
    from_locations = convert_locations(queries['from'], from_type)
    from_size = chunk_size if len(from_locations) > chunk_size else len(from_locations)
    to_locations = convert_locations(queries['to'], to_type)
    to_size = chunk_size if len(to_locations) > chunk_size else len(to_locations)

    data = asyncio.run(async_order_request(from_locations, from_size, to_locations, to_size))
    size = len(data['from']) + len(data['to'])

    print(f'-- {time.time() - start_time} seconds to retrieve {size} orders --')

    data['from'] = remap_orders(data['from'], False if from_type == 'buy' else True)
    data['to'] = remap_orders(data['to'], False if to_type == 'buy' else True)
    data = remove_mismatch_type_ids(data)
    size = len(data['from']) + len(data['to'])

    print(f'-- {time.time() - start_time} seconds to process {size} orders --')

    trades = get_valid_trades(data, sales_tax, min_profit, min_roi, max_budget, max_weight)

    print(f'-- {time.time() - start_time} seconds to get valid {len(trades)} trades --')

    trades = asyncio.run(get_all_jump_counts(trades, route_safety))

    print(f'-- {time.time() - start_time} seconds to apply jump counts to {len(trades)} trades --')

    return {
        'statusCode': 200,
        'body': trades
    }
